File: AlphaGo/Games/Hexagonal.py
import numpy as np
import json


class Hexagonal:
    def __init__(self, board_type, board_size, cell_types):
        self.neighbours = [(-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0)]
        self.board = self._create_board(board_type, board_size, cell_types)
        self.actual_neighbours = {}

    @staticmethod
    def _create_board(board_type, dimensions, cell_types):
        '''

        @param type:
        @param dimensions:
        @param cell_types:
        @return:
        '''

        board = np.array([])
        if board_type == 'triangle':
            board = np.array([[0 if x < dimensions - y else -1 for x in range(dimensions)] for y in range(dimensions)])
        elif board_type == 'diamond':
            board = np.array([[0] * dimensions] * dimensions)
        for num, cell_type in enumerate(cell_types):
            for y, x in cell_type:
                board[y, x] = num if board[y, x] != -1 else -1
        return board

    def get_board(self):
        return self.board

    def set_board(self, board):
        self.board = board

    def get_unoccupied(self):
        return [x * self.board.shape[1] + y for x in range(self.board.shape[0]) for y in
                range(self.board.shape[1]) if self.board[x, y] == 0]

    def get_neighbours(self, pos, value):
        if pos in self.actual_neighbours:
            return self._get_matching_neighbours(self.actual_neighbours[pos], value)
        x, y = (pos // self.board.shape[1], pos % self.board.shape[1])
        neighbours = []
        num_neighbours = []
        for direction in self.neighbours:
            x_next = x + direction[0]
            y_next = y + direction[1]
            if 0 <= x_next < self.board.shape[0] and 0 <= y_next < self.board.shape[1]:
                neighbours.append([x_next, y_next])
                num_neighbours.append(x_next * self.board.shape[1] + y_next)
        self.actual_neighbours[pos] = [neighbours, num_neighbours]
        return self._get_matching_neighbours(self.actual_neighbours[pos], value)

    def _get_matching_neighbours(self, neighbours, value):
        return [neighbours[1][pos] for pos in range(len(neighbours[0])) if self.board[neighbours[0][pos][0], neighbours[0][pos][1]] == value]

    def border(self, pos):
        #
        # 0/\1
        # 2\/3
        x, y = (pos // self.board.shape[1], pos % self.board.shape[1])
        sides = []

        if x == 0:
            sides.append(1)
        if x == self.board.shape[0] - 1:
            sides.append(2)
        if y == 0:
            sides.append(0)
        if y == self.board.shape[1] - 1:
            sides.append(3)
        return sides

    def __getitem__(self, item):
        return self.board[item // self.board.shape[1], item % self.board.shape[1]]

    def __setitem__(self, key, value):
        self.board[key // self.board.shape[1], key % self.board.shape[1]] = value

    def __str__(self):
        """

        :return:
        """

        board_str = ""
        print(self.board.shape)
        for y in range(self.board.shape[0] * 2 - 1):
            board_str += " " * (abs(self.board.shape[0] - y - 1))

            for x in range(y + 1):
                if 0 <= y - x < self.board.shape[0] and x < self.board.shape[0]:
                    if self.board[y - x, x] == -1:
                        return board_str
                    board_str += str(self.board[y - x, x]) + " "
            board_str += "\n"
        return board_str


if __name__ == '__main__':
    print(Hexagonal())

File: AlphaGo/Games/Hexagonal.cpp
#include "Hexagonal.h"

Hexagonal::Hexagonal(int boardSize){
	createBoard(boardSize);
}
Hexagonal::Hexagonal(){
	createBoard(5);
}

void Hexagonal::createBoard(int size){
	board.assign(size, vector<int> (size, 0));
}
vector<vector<int>> Hexagonal::getBoard(){
	return board;
}
void Hexagonal::setBoard(vector<vector<int>> inpBoard){
	board = inpBoard;
}

vector<string>* Hexagonal::getUnoccupied(vector<string>* inpGetUnoccupied){
	inpGetUnoccupied->reserve(board.size()*board[0].size());
	for (int x=0; x < board.size(); x++){
		for (int y=0; y < board[0].size(); y++){
			if (board[x][y] == 0){
				int pos = x * board[0].size() + y;
				(*inpGetUnoccupied).push_back(to_string(pos));
			}
		}
	}
	return inpGetUnoccupied;
}


vector<int>* Hexagonal::getNeighbours(int pos, int value, vector<int>* inpActualNeighbours){
	int x = pos / board[0].size();
	int y = pos % board[0].size();
	inpActualNeighbours->reserve(neighbours.size());
	for (int dir = 0; dir < neighbours.size(); dir++){
		int neigh_x = x + neighbours[dir].first;
		int neigh_y = y + neighbours[dir].second;
		if (neigh_x >= 0 && neigh_x < board.size() &&
			neigh_y >= 0 && neigh_y < board.size() && 
			board[neigh_x][neigh_y] == value){
			(*inpActualNeighbours).push_back(neigh_x * board[0].size() + neigh_y);
		}
	}
	return inpActualNeighbours;
}

pair<int, int> Hexagonal::border(int position){
	/*
	 *	0/\1
	 *	2\/3
	 */
	pair<int, int> borders = make_pair(-1, -1);
	int x = position / board[0].size();
	int y = position % board[0].size();
	//cout << "pos " << position << " " << x << " " << y << endl;

	if (x == 0){
		borders.first = 1;
	}
	else if (x == board.size()-1){
		borders.first = 2;
	}
	if (y == 0){
		borders.second = 0;
	}
	else if (y == board[0].size()-1){
		borders.second = 3;
	}
	//cout << "borders " << borders.first << " " << borders.second << endl;
	return borders;
}

int Hexagonal::getSquare(int pos){
	return board[pos / board[0].size()][pos % board[0].size()];
}


void Hexagonal::setSquare(int pos, int value){
	board[pos / board[0].size()][pos % board[0].size()] = value;
}



/*
int main(){
	int num  = 5;
	cout << "hei" << endl;
	Hexagonal hexGame;
	cout << "Hei" << endl;
};
*/
